### PR TITLE
Tx Encoding and decoding fix

### LINKED ISSUE #23: For the same txid, why is the "status" returned from getTransactionReceipt different between versions 5.5.2 and 6.0.6?
Moved from https://github.com/ethereumclassic/go-ethereum/issues/64

I am trying to upgrade a node from version 5.5.2 to 6.0.6. But I am confused. For a successful transaction, getTransactionReceipt will return "status" to 1 on version 5.5.2. However, in version 6.0.6, "status" is 0 for the same txid.
 Example:
txid: "0x57c6759aa9cd4ab72f60dc0f50f2bd122b00750425f0d7130bd83624424f0837"
on 5.5.2
```json
{
  "blockHash": "0x5f42e57715275fcf44285e94cd187364f3060110e8b77fca0402b2420336b1ce",
  "blockNumber": 8544487,
  "contractAddress": null,
  "cumulativeGasUsed": 21000,
  "from": "0xdf7d7e053933b5cc24372f878c90e62dadad5d42",
  "gasUsed": 21000,
  "logs": [],
  "root": "6e1a89091922a0194173be1ac6159f2425cfda45c0e5d2898e586f4934d13f2f",
  "status": 1,
  "to": "0xb978fba16b86b4406bbed13f7d8719ed85a32cf4",
  "transactionHash": "0x57c6759aa9cd4ab72f60dc0f50f2bd122b00750425f0d7130bd83624424f0837",
  "transactionIndex": 0
}
```
but on 6.0.6
```json
{
  "blockHash": "0x5f42e57715275fcf44285e94cd187364f3060110e8b77fca0402b2420336b1ce",
  "blockNumber": 8544487,
  "contractAddress": null,
  "cumulativeGasUsed": 21000,
  "from": "0xdf7d7e053933b5cc24372f878c90e62dadad5d42",
  "gasUsed": 21000,
  "logs": [],
  "root": "6e1a89091922a0194173be1ac6159f2425cfda45c0e5d2898e586f4934d13f2f",
  "status": 0,
  "to": "0xb978fba16b86b4406bbed13f7d8719ed85a32cf4",
  "transactionHash": "0x57c6759aa9cd4ab72f60dc0f50f2bd122b00750425f0d7130bd83624424f0837",
  "transactionIndex": 0
}
```

### PR DESCRIPTION
**Pull request by austinabell**
Followed status encoding functions from foundation who implicitly ommitted PostState which altered the logic of the functionality of the receipt encoding and decoding, which was not caught by the existing tests or tests pulled from foundation.

Issue was if a Receipt in storage contained the PostState and Status, the Status would not be translated over into the returned value in favour of the PostState (which was removed and irrelevant at the current block).

I made these changes to remain compatible with all previous versions and now stores all fields without ommitting the PostState as was previously intended, but optimizations can be made later. I also added tests to ensure the compatibility and test these edge cases to make sure no changes affect the compatibility.

Please ensure the tx status appears on receipts migrating from a previous version, I cannot test that specifically.

Fixes #23

### CODE REVIEW

**Review by GregTheGreek — DISMISSED** [review]

File: core/types/receipt.go [hunk 1]
```diff
@@ -179,12 +180,12 @@ func (r *ReceiptForStorage) DecodeRLP(s *rlp.Stream) error {
 		return err
 	}
 
-	// Try decoding the receipt without Status first
-	if err := decodeStoredReceiptRLP(r, raw); err == nil {
+	// Try decoding the receipt with Status first
+	if err := decodeStoredReceiptRLPv2(r, raw); err == nil {
```
Comment (r8d8) [inline core/types/receipt.go]: `...v2` or similar suffixes makes a lot of confusion.
It's much easier to have meaningfull domain name, especially when it's not bound to any actuall versioning.

**Review by r8d8 — DISMISSED** [review]

File: core/types/receipt.go (hunk 1 above)
Comment (austinabell) [reply core/types/receipt.go]: Yeah it was just already a long variable name, any suggestions then? Because it isn't tied to any consensus or protocol change either.

Comment (soc1c) [reply core/types/receipt.go]:
```suggestion
	if err := decodeStoredReceiptRLPWithStatus(r, raw); err == nil {
```

length doesn't matter, readability does

Comment (austinabell) [reply core/types/receipt.go]: ok I'll do that now then if no other changes yet

**Review by GregTheGreek — APPROVED** [review]

**Review by tzdybal — APPROVED** [review]
LGTM

**Review by noot — APPROVED** [review]
